.github/cloudbuild/android.yaml

```diff
@@ -0,0 +1,111 @@
+steps:
+# We need this step to correctly identify dockerfile tag
+- id: fetch-repo-history
+  name: gcr.io/cloud-builders/git
+  timeout: 10s
+  args:
+  - fetch
+  - --unshallow
+# Download DOCKERFILE_COMMIT tag if it exists to skip docker image generation,
+#   or download the :latest tag and use it as a cache,
+#   or skip downloading if :latest doesn't exist yet.
+# This being a separate step helps readability in Cloud Build console.
+- id: cache-old-image
+  name: gcr.io/cloud-builders/docker
+  timeout: 600s # 10 minutes
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)
+    docker pull gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT \
+    || docker pull gcr.io/$PROJECT_ID/mlperf-mobile-android:latest \
+    || true
+# This step overrides the :latest tag of the image, so that we can use it in later steps.
+- id: build-new-image
+  name: gcr.io/cloud-builders/docker
+  timeout: 1800s # 30 minutes
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)
+    docker build \
+      -t gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT \
+      -t gcr.io/$PROJECT_ID/mlperf-mobile-android:latest \
+      --cache-from gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT \
+      --cache-from gcr.io/$PROJECT_ID/mlperf-mobile-android:latest \
+      -f android/docker/mlperf_mobile/Dockerfile \
+      .
+# If the build fails artifacts are not uploaded automatically, so we save them manually before build
+- id: push-new-image
+  name: gcr.io/cloud-builders/docker
+  timeout: 600s # 10 minutes
+  entrypoint: bash
+  args:
+  - -xc
+  - |
+    DOCKERFILE_COMMIT=$$(git log -n 1 --pretty=format:%H -- android/docker/mlperf_mobile/Dockerfile)
+    docker push gcr.io/$PROJECT_ID/mlperf-mobile-android:$$DOCKERFILE_COMMIT
+    docker push gcr.io/$PROJECT_ID/mlperf-mobile-android:latest
+- id: build-apks
+  name: gcr.io/$PROJECT_ID/mlperf-mobile-android:latest
+  waitFor:
+  - build-new-image # we can run this step in parallel with image uploading
+  timeout: 10800s # 3 hours
+  entrypoint: bash
+  args:
+  - -xc

```

Comment:
This is just my preference, so it's not important. I like to define all the commands in a `Makefile` then execute it in a build file by invoking `make step_abc`. This make it easier to change CI/CD tool / service or use multiple ones. And using`Makefile` has some advantage because it is supported by almost every IDE.

Reply:
I agree, this would be more convenient, and we would automatically update the CI when we update the makefile we use locally.
This could also unify local builds with docker, local builds without docker, and builds on CI.

Unfortunately, right now I don't see appropriate command in our makefiles. We should refactor them.
@mcharleb is reworking the makefiles right now, maybe he could help here. Or else we should refactor the makefiles later.
For this PR I suggest just leaving the build command as is.